src/vfio_device.rs

```diff
@@ -0,0 +1,962 @@
+            debug!("\toffset 0x{:x}", region.offset);
+
+            regions.push(region);
+        }
+
+        Ok(regions)
+    }
+}
+
+/// Vfio device to access underline hardware devices.
+///
+/// The VFIO device API includes ioctls for describing the device, the I/O regions and their
+/// read/write/mmap offsets on the device descriptor, as well as mechanisms for describing and
+/// registering interrupt notifications.
+pub struct VfioDevice {
+    device: ManuallyDrop<File>,
+    flags: u32,
+    regions: Vec<VfioRegion>,
+    irqs: HashMap<u32, VfioIrq>,
+    group: Arc<VfioGroup>,
+    container: Arc<VfioContainer>,
+}
+
+impl VfioDevice {
+    /// Create a new vfio device, then guest read/write on this device could be transferred into kernel vfio.
+    ///
+    /// # Parameters
+    /// * `sysfspath`: specify the vfio device path in sys file system.
+    /// * `container`: the new VFIO device object will bind to this container object.
+    pub fn new(sysfspath: &Path, container: Arc<VfioContainer>) -> Result<Self> {
+        let uuid_path: PathBuf = [sysfspath, Path::new("iommu_group")].iter().collect();
+        let group_path = uuid_path.read_link().map_err(|_| VfioError::InvalidPath)?;
+        let group_osstr = group_path.file_name().ok_or(VfioError::InvalidPath)?;
+        let group_str = group_osstr.to_str().ok_or(VfioError::InvalidPath)?;
+        let group_id = group_str
+            .parse::<u32>()
+            .map_err(|_| VfioError::InvalidPath)?;
+
+        let group = container.get_group(group_id)?;
+        let device_info = group.get_device(sysfspath)?;
+        let regions = device_info.get_regions()?;
+        let irqs = device_info.get_irqs()?;
+
+        Ok(VfioDevice {
+            device: ManuallyDrop::new(device_info.device),
+            flags: device_info.flags,
+            regions,
+            irqs,
+            group,
+            container,
+        })
+    }
+
+    /// VFIO device reset only if the device supports being reset.
+    pub fn reset(&self) {
+        if self.flags & VFIO_DEVICE_FLAGS_RESET != 0 {
+            unsafe { ioctl(self, VFIO_DEVICE_RESET()) };
+        }
+    }
+
+    /// Get information about VFIO IRQs.
+    ///
+    /// # Arguments
+    /// * `irq_index` - The type (INTX, MSI or MSI-X) of interrupts to enable.
+    pub fn get_irq_info(&self, irq_index: u32) -> Option<&VfioIrq> {
+        self.irqs.get(&irq_index)
+    }
+
+    /// Trigger a VFIO device IRQ from userspace.
+    ///
+    /// Once a signaling mechanism is set, DATA_BOOL or DATA_NONE can be used with ACTION_TRIGGER
+    /// to perform kernel level interrupt loopback testing from userspace (ie. simulate hardware
+    /// triggering).
+    ///
+    /// # Arguments
+    /// * `irq_index` - The type (INTX, MSI or MSI-X) of interrupts to enable.
+    /// * `vector` - The sub-index into the interrupt group of `irq_index`.
+    pub fn trigger_irq(&self, irq_index: u32, vector: u32) -> Result<()> {
```

Comment:
Shouldn't `irq_index` actually be `irq_type`? This name exposes implementation details of how vfio works in the kernel and this should not concern the user of this API.

Reply:
This is to keep the same terms as VFIO docs/header files.